### ISSUE TITLE
Cannot drop PRIMARY key on partition

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#3750]
Steps to reproduce:

1. Create table, and a primary key:

```
postgres=# CREATE TABLE mytable (id int4) DISTRIBUTED BY (id) PARTITION BY LIST (id) (PARTITION p1 VALUES (1)) ;
NOTICE:  CREATE TABLE will create partition "mytable_1_prt_p1" for table "mytable"
CREATE TABLE
postgres=# ALTER TABLE mytable ADD primary key (id);
NOTICE:  ALTER TABLE / ADD PRIMARY KEY will create implicit index "mytable_pkey" for table "mytable"
NOTICE:  ALTER TABLE / ADD PRIMARY KEY will create implicit index "mytable_1_prt_p1_pkey" for table "mytable_1_prt_p1"
ALTER TABLE
```

2. Drop the primary key:
```
postgres=# alter table mytable drop constraint mytable_pkey;
ALTER TABLE
```

At this point, I would've expected it to drop the primary key on the partition as well. But it did not. As evidenced by this:
```
postgres=# ALTER TABLE mytable ADD primary key (id);
NOTICE:  ALTER TABLE / ADD PRIMARY KEY will create implicit index "mytable_pkey" for table "mytable"
ERROR:  multiple primary keys for table "mytable_1_prt_p1" are not allowed
```

To make things worse, the primary key on the partition cannot be dropped:
```
postgres=# alter table mytable_1_prt_p1 drop constraint mytable_1_prt_p1_pkey;
ERROR:  can't drop a constraint from "mytable_1_prt_p1"; it is part of a partitioned table
HINT:  You may be able to perform the operation on the partitioned table as a whole.
```

**Comment by guofengrichard** [comment]
The primary key on partition cannot be dropped because the name of the constraint does not match the name in the command line. For example, suppose we have table `mytable` and its child table `mytable_1_prt_p1`. If we add primary key for `mytable`, the primary constraints would be `mytable_pkey` and `mytable_1_prt_p1_pkey`, respectively. When we are trying to drop the primary key with `alter table mytable drop constraint mytable_pkey;`, the constraint of the parent table is dropped with no problem. But when it comes to the child table, it figures out that the name of its primary constraint is not the same as in the command line, so it would refuse to perform the deletion.

**Comment by guofengrichard** [comment]
I don't think it is a good idea to check the constraint's name, because the name can be changed casually:
```
gpadmin=# alter table mytable_1_prt_p1 rename constraint mytable_1_prt_p1_pkey to mytable_1_prt_p1_randomname;
ALTER TABLE
```
Therefore, it is difficult to compose the right name for the partition's constraint, as does in `ATPrepDropConstraint()`. 

PostgreSQL 10 or above does the constraint deletion for partitioned table via dependency mechanism, which is a better way I think we should follow. I am wondering if we should backport that logic from PostgreSQL (not sure if it is easy), or just leave this issue as a known issue and wait util we merge up to PostgreSQL 10.

**Comment by kalensk** [comment]
This has been fixed with #7047.

Verified using the original reproduction steps as outlined below. However, we are no longer displaying the following notice's:
```
NOTICE:  ALTER TABLE / ADD PRIMARY KEY will create implicit index "mytable_pkey" for table "mytable"
NOTICE:  ALTER TABLE / ADD PRIMARY KEY will create implicit index "mytable_1_prt_p1_pkey" for table "mytable_1_prt_p1"
```

**Verification of Reproduction Steps:**
1. Create table, and a primary key:
```
postgres=# CREATE TABLE mytable (id int4) DISTRIBUTED BY (id) PARTITION BY LIST (id) (PARTITION p1 VALUES (1)) ;
NOTICE:  CREATE TABLE will create partition "mytable_1_prt_p1" for table "mytable"
CREATE TABLE
postgres=# ALTER TABLE mytable ADD primary key (id);
ALTER TABLE
```

2. Drop the primary key (which now cascades to the partition)
```
postgres=# alter table mytable drop constraint mytable_pkey;
ALTER TABLE

postgres=# \d+ mytable_1_prt_p1
                   Table "public.mytable_1_prt_p1"
 Column |  Type   | Modifiers | Storage | Stats target | Description
--------+---------+-----------+---------+--------------+-------------
 id     | integer | not null  | plain   |              |
Check constraints:
    "mytable_1_prt_p1_check" CHECK (id = 1)
Inherits: mytable
Distributed by: (id)

postgres=# \d+ mytable
                       Table "public.mytable"
 Column |  Type   | Modifiers | Storage | Stats target | Description
--------+---------+-----------+---------+--------------+-------------
 id     | integer | not null  | plain   |              |
Child tables: mytable_1_prt_p1
Distributed by: (id)
Partition by: (id)

```

3. Adding the primary key again succeeds
```
postgres=# ALTER TABLE mytable ADD primary key (id);
ALTER TABLE

postgres=# \d+ mytable
                       Table "public.mytable"
 Column |  Type   | Modifiers | Storage | Stats target | Description
--------+---------+-----------+---------+--------------+-------------
 id     | integer | not null  | plain   |              |
Indexes:
    "mytable_pkey" PRIMARY KEY, btree (id)
Child tables: mytable_1_prt_p1
Distributed by: (id)
Partition by: (id)

postgres=# \d+ mytable_1_prt_p1
                   Table "public.mytable_1_prt_p1"
 Column |  Type   | Modifiers | Storage | Stats target | Description
--------+---------+-----------+---------+--------------+-------------
 id     | integer | not null  | plain   |              |
Indexes:
    "mytable_1_prt_p1_pkey" PRIMARY KEY, btree (id)
Check constraints:
    "mytable_1_prt_p1_check" CHECK (id = 1)
Inherits: mytable
Distributed by: (id)
```

4. Attempting to drop the re-added primary key is only allowed when dropping from the root.
```
postgres=# alter table mytable_1_prt_p1 drop constraint mytable_1_prt_p1_pkey;
ERROR:  cannot drop inherited constraint "mytable_1_prt_p1_pkey" of relation "mytable_1_prt_p1"

postgres=# alter table mytable drop constraint mytable_pkey;
ALTER TABLE

postgres=# \d+ mytable
                       Table "public.mytable"
 Column |  Type   | Modifiers | Storage | Stats target | Description
--------+---------+-----------+---------+--------------+-------------
 id     | integer | not null  | plain   |              |
Child tables: mytable_1_prt_p1
Distributed by: (id)
Partition by: (id)

postgres=# \d+ mytable_1_prt_p1
                   Table "public.mytable_1_prt_p1"
 Column |  Type   | Modifiers | Storage | Stats target | Description
--------+---------+-----------+---------+--------------+-------------
 id     | integer | not null  | plain   |              |
Check constraints:
    "mytable_1_prt_p1_check" CHECK (id = 1)
Inherits: mytable
Distributed by: (id)
```
